File: tests/memory_benchmark.py
import os
import base64
import psutil
import time

def get_memory_usage():
    process = psutil.Process(os.getpid())
    return process.memory_info().rss

def benchmark_legacy(file_path):
    # Simulate PHP's file_get_contents + base64_encode + payload construction
    start_mem = get_memory_usage()

    with open(file_path, 'rb') as f:
        file_content = f.read()

    # PHP's base64_encode returns a string
    base64_file = base64.b64encode(file_content)

    payload = b""
    boundary = b"--------------------------" + str(time.time()).encode()

    payload += b"--" + boundary + b"\r\n"
    payload += b'Content-Disposition: form-data; name="file"' + b"\r\n\r\n"
    payload += base64_file + b"\r\n"
    payload += b"--" + boundary + b"--\r\n"

    end_mem = get_memory_usage()

    # Hold onto payload so it's not GC'd before measurement
    len(payload)

    return end_mem - start_mem

def benchmark_optimized(file_path):
    # Simulate usage of a file handle / CurlFile
    start_mem = get_memory_usage()

    # In the optimized version, we don't read the whole file.
    # We just open a file object or pass the path.
    # Emulating passing a file handle/path
    f = open(file_path, 'rb')

    post_fields = {
        'file': f,
        'fileName': os.path.basename(file_path),
        'useUniqueFileName': 'false'
    }

    end_mem = get_memory_usage()
    f.close()

    return end_mem - start_mem

def main():
    file_path = "temp_large_file.dat"
    # Create 20MB file
    size_mb = 20
    with open(file_path, 'wb') as f:
        f.write(os.urandom(size_mb * 1024 * 1024))

    print(f"Benchmark with {size_mb}MB file:")

    try:
        mem_diff_legacy = benchmark_legacy(file_path)
        print(f"Legacy Implementation Memory Increase: {mem_diff_legacy / 1024 / 1024:.2f} MB")

        mem_diff_optimized = benchmark_optimized(file_path)
        print(f"Optimized Implementation Memory Increase: {mem_diff_optimized / 1024 / 1024:.2f} MB")

    finally:
        if os.path.exists(file_path):
            os.remove(file_path)

if __name__ == "__main__":
    main()

File: tests/memory_benchmark.php
<?php

// memory_benchmark.php

function benchmark_current_implementation($filePath) {
    $startMemory = memory_get_peak_usage(true);

    // Simulate the logic in ImageKitAdapter::store
    if (!file_exists($filePath)) {
        throw new Exception("File not found");
    }

    $fileContent = file_get_contents($filePath);
    if ($fileContent === false) {
        throw new Exception("Failed to read file");
    }

    // Use base64 for reliable transmission via WP remote post
    $base64File = base64_encode($fileContent);

    $payload = '';
    $boundary = '--------------------------' . microtime(true);

    // file
    $payload .= "--" . $boundary . "\r\n";
    $payload .= 'Content-Disposition: form-data; name="file"' . "\r\n\r\n";
    $payload .= $base64File . "\r\n";

    // ... other fields would be added here ...

    $payload .= "--" . $boundary . "--\r\n";

    // Simulate passing to wp_remote_post (which would keep it in memory)
    $args = ['body' => $payload];

    $endMemory = memory_get_peak_usage(true);

    return $endMemory - $startMemory;
}

function benchmark_optimized_implementation($filePath) {
    $startMemory = memory_get_peak_usage(true);

    // Optimized approach using CURLFile (simulation)
    if (!file_exists($filePath)) {
        throw new Exception("File not found");
    }

    // We don't read the file content into a variable.
    // We create a CURLFile object.
    $cFile = new CURLFile($filePath);

    $postFields = [
        'file' => $cFile,
        'fileName' => basename($filePath),
        'useUniqueFileName' => 'false',
    ];

    // In real implementation, we would pass this to curl_setopt(..., CURLOPT_POSTFIELDS, $postFields)
    // The memory usage should be minimal as the file is not read into PHP memory.

    $endMemory = memory_get_peak_usage(true);

    return $endMemory - $startMemory;
}

// Create a dummy file (e.g. 5MB)
$tempFile = __DIR__ . '/temp_large_file.dat';
$size = 5 * 1024 * 1024; // 5MB
$fp = fopen($tempFile, 'w');
// Write in chunks to avoid memory spike during creation
$chunk = str_repeat('0', 1024 * 1024);
for ($i = 0; $i < 5; $i++) {
    fwrite($fp, $chunk);
}
fclose($fp);

echo "Benchmark with " . ($size / 1024 / 1024) . "MB file:\n";

try {
    // Force GC
    gc_collect_cycles();
    $memUsageLegacy = benchmark_current_implementation($tempFile);
    echo "Legacy Implementation Memory Usage: " . number_format($memUsageLegacy / 1024 / 1024, 2) . " MB\n";

    // Force GC
    gc_collect_cycles();
    $memUsageOptimized = benchmark_optimized_implementation($tempFile);
    echo "Optimized Implementation Memory Usage: " . number_format($memUsageOptimized / 1024 / 1024, 2) . " MB\n";

} catch (Exception $e) {
    echo "Error: " . $e->getMessage() . "\n";
}

// Cleanup
if (file_exists($tempFile)) {
    unlink($tempFile);
}
